plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/building/CloseProjectTaskScheduler.java

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright (c) 2018 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ui.building;
+
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.resources.WorkspaceJob;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.SubMonitor;
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.emf.ecore.resource.impl.ResourceSetImpl;
+import org.eclipse.n4js.internal.RaceDetectionHelper;
+import org.eclipse.n4js.ui.building.ClosedProjectQueue.Task;
+import org.eclipse.xtext.builder.impl.BuildData;
+import org.eclipse.xtext.builder.impl.Messages;
+import org.eclipse.xtext.builder.impl.ProjectOpenedOrClosedListener;
+import org.eclipse.xtext.builder.impl.QueuedBuildData;
+import org.eclipse.xtext.builder.impl.ToBeBuilt;
+import org.eclipse.xtext.resource.impl.ResourceDescriptionsProvider;
+import org.eclipse.xtext.ui.shared.internal.ListenerRegistrar;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * Specialized project open or closed listener that fixes a bunch of races and misbehavior that is present in the
+ * default impl.
+ */
+@SuppressWarnings("restriction")
+@Singleton
+public class CloseProjectTaskScheduler extends ProjectOpenedOrClosedListener {
+
+	private final class RemoveProjectJob extends WorkspaceJob {
```

Comment:
is this job cancellable? should it be cancellable?

Reply:
It passes the progress monitor into the builder. So yes, it will be cancelled if the monitor is cancelled.

Reply:
I was just wondering if we want this to be cancellable. Cancelling cleanup triggered by closing project leaves workspace / index in weird state.
On the other hand any build user cancels leaves IDE in weird state. So I guess it is ok.